### ISSUE TITLE
Fix performance tests failing when migrating to `useOnyx()`

### ISSUE THREAD

**Issue by lakchote** [Expensify/App#51562]
In https://github.com/Expensify/App/pull/50544 issue , migrating from `withOnyx()` to `useOnyx()` caused the performance tests to fail (see discussion [here](https://github.com/Expensify/App/pull/50544#discussion_r1815597972).

Investigate why and fix it in order to be able to use `useOnyx()` and solve the lint issue.

**Comment by lakchote** [comment]
@blazejkustra or @fabioh8010 which one of you two would like to work on this?

**Comment by blazejkustra** [comment]
I can look into this 👀

**Comment by rojiphil** [comment]
@lakchote Do you need a C+ here? If needed, I can help as I have the context.

**Comment by lakchote** [comment]
@rojiphil yes I'm assigning you as the C+ reviewer, as you have context on the underlying issue.

**Comment by blazejkustra** [comment]
Update: I started investigating yesterday ([WIP PR](https://github.com/Expensify/App/pull/51573)). It looks like the only hook that increases re-renders is the modal one:
```ts
const [modal] = useOnyx(ONYXKEYS.MODAL);
```

I tried moving this hook to a wrapper, checking if value is loaded but always with the same result - more renders and failing perf tests. That's why I started digging deeper in `react-native-onyx` and `reassure` libraries. 

For now my findings are that modal hook make the component re-render even when value is left unchanged. This is the reason why we see this weird pattern in the app (it looks like a perfect example where `useOnyx(ONYXKEYS.MODAL)`, should be used):
```ts
    const [modal, setModal] = useState<Modal>({
        willAlertModalBecomeVisible: false,
        isVisible: false,
    });

    useEffect(() => {
        const unsubscribeOnyxModal = onyxSubscribe({
            key: ONYXKEYS.MODAL,
            callback: (modalArg) => {
                if (modalArg === undefined) {
                    return;
                }
                setModal(modalArg);
            },
        });
        return () => {
            unsubscribeOnyxModal();
        };
    }, []);
```

My guess is that there is a problem in `react-native-onyx` implementation but I'll keep digging.

Maybe you have some additional context that you'd like to share? @rojiphil @daledah (as you introduced `onyxSubscribe` in `EdicationalTooltip`)

**Comment by blazejkustra** [comment]
Update: 
Internally `useOnyx` uses `useSyncExternalStore`, and there is a [known issue with `useSyncExternalStore`](https://github.com/facebook/react/issues/24831) that will be fixed in React 19. Basically `useSyncExternalStore` purposely tries to render as quick as possible, sometimes it won’t batch the updates - this means that a slower `withOnyx` might be causing less re-renders (and same when using `onyxSubscribe` as it doesn't use `useSyncExternalStore`)



From my research I observed that there are a lot of frequent updates to the `ONYXKEYS.MODAL` key, meaning more renders caused by `useOnyx`. However that's only the case in performance tests, in the app I don't see any differences in re-renders which is making it really hard to debug.

<details>
<summary>useOnyx</summary>

https://github.com/user-attachments/assets/63012629-dca4-4acb-981a-22d9261f2558

</details>

<details>
<summary>withOnyx</summary>

https://github.com/user-attachments/assets/6fe13895-2487-4936-bb31-0647decec0f2

</details>

My proposal is to ignore failing perf tests and accept more re-renders, my reasoning it that in real environment there is no difference in render count.

**Comment by lakchote** [comment]
So when the [known issue with `useSyncExternalStore`](https://github.com/facebook/react/issues/24831) is solved with React 19, we shouldn't have the problem anymore in our performance tests, right?

Does this mean we'll may encounter the same problem down the road with faulty performance tests that are not truly indicative of a re-rendering performance issue until then? If yes, I think we should get more eyes on this by posting in #expensify-open-source.

**Comment by blazejkustra** [comment]
> so when the https://github.com/facebook/react/issues/24831 is solved with React 19, we shouldn't have the problem anymore in our performance tests, right?

Yes, that's my assumption. If these onyx updates are triggered in a short period of time React is going to batch them, resulting in less renders.

> Does this mean we'll may encounter the same problem down the road with faulty performance tests that are not truly indicative of a re-rendering performance issue

It's more complicated than that, performance tests try to recreate the real environment as best as possible, but we can't avoid mocking a bunch of libraries, providers and components. It usually doesn't make a big difference, but I'm sure we have some tests that are flaky or doesn't bring much value. There is even a proposal that aims to fix [this](https://expensify.slack.com/archives/C05LX9D6E07/p1729767940018789), and [issue](https://github.com/Expensify/App/issues/51637).

**Comment by blazejkustra** [comment]
On top of that, the failing test will be removed as a result of [this](https://github.com/Expensify/App/pull/51461/files#diff-a1b8a91a50ff890b57a89cc7403bb6f0c29835267735fe5908f3f4fd6e2f8e0d) PR 😅

**Comment by rojiphil** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [x] 1z. Other:
This is part of the migration work from `withOnyx` to `useOnyx`

Where bug was reported:
  - [ ] 2a. Reported on production
  - [ ] 2b. Reported on staging (deploy blocker)
  - [x] 2c. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [x] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    [Link to comment](https://github.com/Expensify/App/pull/50544#issuecomment-2441465871):

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: Not applicable here.

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

As this was migration work, we can skip this step as no new test case is introduced here. 

Do we agree 👍 or 👎

**Comment by sakluger** [comment]
Summarizing payment on this issue:

Contributor: @blazejkustra - No payment required, contractor
Contributor+: @  $250, Sent offer via Upwork: https://www.upwork.com/nx/wm/offer/104853498

**Comment by rojiphil** [comment]
@sakluger Accepted the offer. Thanks.

**Comment by sakluger** [comment]
All paid, thanks!
